File: SOLUTIONS/LEETCODE/linkedLists/getIntersectionNode/getIntersectionNode.py
def getIntersectionNode (headA, headB):
  # check if either list is empty
  if not headA or not headB : return None
  
  pointerA = headA
  pointerB = headB
  
  # traverse both lists with dual pointers
  while pointerA is not pointerB:
    # switch to the next node or the other list's head
    pointerA = pointerA.next if pointerA else headB
    pointerB = pointerB.next if pointerB else headA
    
  return pointerA

File: SOLUTIONS/LEETCODE/EASY/linkedLists/getIntersectionNode/getIntersectionNode.js
function getIntersectionNode (headA, headB) {
  if (!headA || !headB) {return null} 

  //
  let pointerA = headA
  let pointerB = headB

  //
  while (pointerA !== pointerB) {
    // when pointerA reaches to the end of listA, let it redirect to headB
    pointerA = pointerA ? pointerA.next : headB
    // when pointerB reaches to the end of listB, let it redirect to headA
    pointerB = pointerB ? pointerB.next : headA
  }
  return pointerA
}